lowball_ad_auth_provider/auth_provider.py

```diff
@@ -0,0 +1,268 @@
+
+        conn = Connection(server=self.get_server(),
+                          user=self.domain + "\\" + auth_package.username,
+                          password=auth_package.password,
+                          authentication=NTLM)
+
+        # Connects with user; True if Valid Creds and Server Reachable
+        if conn.bind():
+            if conn.search(
+                    search_base=self.base_dn,
+                    search_filter=f'(sAMAccountName={auth_package.username})',
+                    attributes="memberOf"):
+                user_data = json.loads(conn.response_to_json())
+                user_groups = user_data['entries'][0]['attributes']['memberOf']
+
+                roles = self.get_roles(user_groups)
+                conn.unbind()
+                return ClientData(client_id=auth_package.username, roles=roles)
+
+            # We were able to bind but the user wasn't found. Likely a config issue with the base DN
+            else:
+                conn.unbind()
+                exception = AuthenticationNotInitializedException()
+                exception.description = "Unable to locate user though authentication succeeded. Check configuration"
+
+                raise exception
+        else:
+            raise InvalidCredentialsException
+
+    @property
+    def auth_package_class(self):
+        """The auth package class that this class' `authenticate` method accepts."""
+        return ADAuthPackage
+
+    def get_client(self, client_id):
+        """if service_account is configured, will enable users to create their own tokens
+        The service account will need to have permissions to search for the clients
+
+        :param client_id:
+        :return:
+        """
+        if not self._service_account:
+            exception = AuthenticationNotInitializedException()
+            exception.description = "get_client not configured. Must include service_account in service configuration"
+            raise exception
+        else:
+            if any(c in client_id for c in INVALID_SAMACCOUNTNAME_CHARS):
+                raise BadRequestException("Submitted samaccountname contained invalid characters")
+
+            conn = Connection(server=self.get_server(),
+                              user=self.domain + "\\" + self.service_account,
+                              password=self._service_account_password,
+                              authentication=NTLM)
+            if conn.bind():
+                if conn.search(
+                        search_base=self.base_dn,
+                        search_filter=f'(sAMAccountName={client_id})',
+                        attributes="memberOf"):
+                    user_data = json.loads(conn.response_to_json())
+                    user_groups = user_data['entries'][0]['attributes']['memberOf']
+                    roles = self.get_roles(user_groups)
+                    conn.unbind()
```

Comment:
```suggestion
                    conn.unbind()
                    user_groups = user_data['entries'][0]['attributes']['memberOf']
                    roles = self.get_roles(user_groups)

```
Same suggestion as before